### ISSUE TITLE
Make DateTimeRange hashable

### ISSUE THREAD

**Issue by gdalmau** [thombashi/DateTimeRange#39]
Implement `__hash__` function on `DateTimeRange` to be able to use it, for example, in combination with the decorator `@lru_cache`.

**Comment by thombashi** [comment]
Thank you for your feedback.

> Implement __hash__ function on DateTimeRange

`DateTimeRange` is a mutable class. According to the official document, such a class should not implement `__hash__()`
Therefore, may require a new immutable class like `FrozenDateTimeRange`.

**Comment by gdalmau** [comment]
> `DateTimeRange` is a mutable class. According to the official document, such a class should not implement `__hash__()`.

Good point. Since `DateTimeRange` is composed of two `datetime` objects, which are not mutable, might it be an idea to also make `DateTimeRange` immutable?

**Comment by thombashi** [comment]
> DateTimeRange is composed of two datetime objects, which are immutable,

This is true. However, I believe `DateTimeRange` is a mutable class.

Implementation of `__hash__` method for `DateTimeRange` itself is pretty easy. for example:

    def __hash__(self):
        return hash((self.start_datetime, self.end_datetime))

This hash output will be change if `start_datetime`/`end_datetime` value changes:

```py
from datetimerange import DateTimeRange

time_range = DateTimeRange.from_range_text("2015-03-22T10:00:00+0900 - 2015-03-22T10:10:00+0900")
print(hash(time_range))
time_range.set_end_datetime("2016-03-22T10:10:00+0900")
print(hash(time_range))
```

```
-2291305032593123401
5944198212632818997
```

**Comment by gdalmau** [comment]
> This hash output will be change if `start_datetime`/`end_datetime` value changes

I understand, that's why I proposed the idea to make `DateTimeRange` immutable. For that, `set_start_datetime` and `set_end_datetime` and any other setter should return another instance, similar to what [`datetime.replace()`](https://docs.python.org/3/library/datetime.html#datetime.date.replace) does. If that sounds like an idea, how would you prefer to approach it? We can make use of [`__slots__`](https://docs.python.org/3/reference/datamodel.html#slots) or [`dataclass`](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass) with `frozen=True`.

**Comment by thombashi** [comment]
I still would prefer to approach that creates an immutable `DateTimeRange` class as a different class.

> set_start_datetime and set_end_datetime and any other setter should return another instance, similar to what datetime.replace() does. If that sounds like an idea, how would you prefer to approach it? We can make use of __slots__ or dataclass with frozen=True.

because these approaches will break the backward compatibility of `DateTimeRange` class.
